keps/sig-node/5328-node-capabilities/README.md

```diff
@@ -0,0 +1,950 @@
+
+
+## Existing Mechanisms and Limitations
+
+The Kubernetes scheduler currently uses two primary mechanisms to control pod placement onto specific nodes:
+
+1. Taints and Tolerations 
+
+    Primarily used to **restrict** which pods can schedule onto specific nodes. Commonly used to manage specialized hardware resources. 
+
+    **Standard Usage Pattern:**
+*   Cluster Administrators apply taints to nodes equipped with special capabilities; cloud providers may also automate this tainting for certain NodePool configurations.
+*   Developers add corresponding tolerations in their Pod specifications for workloads to be able to run on these nodes. Alternatively, cluster administrator or cloud provider could also inject tolerations through admission webhooks. 
+
+2. Node Labels and Node Selectors/Affinity
+
+    Primarily used to **attract** specific nodes for pods based on the node's characteristics. By applying specific labels to nodes (reflecting Kubelet features, OS version, etc.), we can enable pods to use selectors or affinity to ensure they run on specific nodes.
+
+    **Standard Usage Pattern:**
+*   Cluster Administrators apply specific Labels to nodes to indicate the presence of certain features. This involves applying descriptive labels (e.g., `kubelet.config.k8s.io/some-alpha-feature=enabled`, `node.kubernetes.io/gvisor-enabled=true`)
+*   Optionally, for well-defined features may need to create other resources (e.g., [RunTimeClass](https://kubernetes.io/docs/concepts/containers/runtime-class/)) that bundle these feature requirements with a node selector targeting the corresponding node labels.
+*   Developers specify their workload's dependency on these features in the PodSpec either directly (spec.nodeSelector) or other abstractions enabling the scheduler to match them to capable nodes.
+
+**Drawbacks**:
+
+1. Operational Overhead: Cluster administrators have to add the necessary taints/labels to nodes as indirect signals of features and resources, and workloads should use corresponding tolerations/selectors. This needs to be done manually or automations built (webhooks, controllers) to handle standard usage patterns. 
+2. Scheduling constraints are encoded indirectly rather than being implicitly understood by kube-scheduler. 
+3. Incorrect configurations can lead to scheduling failures or suboptimal placements.
+
+## Proposal
+
+This proposal introduces a new field `Capabilities` to `Node.Status` to expose information which the kube-scheduler and/or the admission controller would use to make more informed decisions. The Kubelet is primarily responsible for discovering, consolidating, and reporting `Capabilities` to the API server.
+
+**Node Capabilities Requirements:**
```

Comment:
will kubelet keep reconciling the capabilities, or it will register them on startup and forget? In other words, if some external controller removed or added a new capability, will kubelet keep trying to return it back to the state it believes it is?

Reply:
Kubelet will build this list of the capabilities once during startup and uses this the subsequent node status updates. If there is an external change to capabilities, the periodic Node Status patch update should reconcile it back to the version that kubelet has (in [tryUpdateNodeStatus](https://github.com/kubernetes/kubernetes/blob/9da4b7ecfb60ea484f8e6c89c42e63506571bc68/pkg/kubelet/kubelet_node_status.go#L496))

Reply:
Can you add a note about reconciliation to the text please

Reply:
Done. Added in "Kubelet Changes" section.